### PR TITLE
feat: bidirectional streaming source

### PR DESCRIPTION
**Pull request by yhl25**

### CODE REVIEW

File: examples/src/test/java/io/numaproj/numaflow/examples/server/ServerTest.java [hunk 1]
```diff
@@ -202,41 +203,42 @@ public void testSinkServerInvocation() {
...
-        }
-
-        try {
-            sourcerClient.close();
-            sourcerTestKit.stopServer();
-        } catch (InterruptedException e) {
-            Assertions.fail("Failed to stop server");
-        }
-    }
+// FIXME: once tester kit changes are done for bidirectional streaming source
```
Comment (vigith) [inline examples/src/test/java/io/numaproj/numaflow/examples/server/ServerTest.java:206]: Can you open an issue in numaflow repo to track this?

File: examples/src/main/java/io/numaproj/numaflow/examples/source/simple/SimpleSource.java [hunk 2]
```diff
@@ -70,16 +72,15 @@ public void read(ReadRequest request, OutputObserver observer) {
...
         // remove the acknowledged messages from the map
-        for (Offset offset : request.getOffsets()) {
-            messages.remove(Longs.fromByteArray(offset.getValue()));
-        }
+        messages.remove(offset);
     }
 
     @Override
     public long getPending() {
         // pending messages will be zero for a simple source
```
Comment (KeranYang) [inline examples/src/main/java/io/numaproj/numaflow/examples/source/simple/SimpleSource.java:82]: Please update the comment as well.

File: examples/src/test/java/io/numaproj/numaflow/examples/source/simple/SimpleSourceTest.java [hunk 3]
```diff
@@ -58,5 +63,14 @@ public void testPending() {
...
 
+    @Test
+    public void testLong() {
+        Long l = 1L;
+        byte[] bytes = Longs.toByteArray(l);
+
+        Long x = Longs.fromByteArray(bytes);
+        Assertions.assertEquals(l, x);
+    }
+
```
Comment (KeranYang) [inline examples/src/test/java/io/numaproj/numaflow/examples/source/simple/SimpleSourceTest.java:74]: We don't need to unit test a library. Please remove it.

File: examples/src/test/java/io/numaproj/numaflow/examples/source/simple/SimpleSourceTest.java [hunk 4]
```diff
@@ -1,8 +1,10 @@
 package io.numaproj.numaflow.examples.source.simple;
 
+import com.google.common.primitives.Longs;
 import io.numaproj.numaflow.sourcer.Message;
 import io.numaproj.numaflow.sourcer.Offset;
 import io.numaproj.numaflow.sourcer.SourcerTestKit;
+import org.junit.Ignore;
```
Comment (KeranYang) [inline examples/src/test/java/io/numaproj/numaflow/examples/source/simple/SimpleSourceTest.java:7]: is this import being used?

File: examples/src/test/java/io/numaproj/numaflow/examples/source/simple/SimpleSourceTest.java [hunk 5]
```diff
@@ -38,9 +40,12 @@ public void test_ReadAndAck() {
         for (Message message : testObserver.getMessages()) {
             offsets.add(message.getOffset());
         }
-        SourcerTestKit.TestAckRequest ackRequest = SourcerTestKit.TestAckRequest.builder()
-                .offsets(offsets).build();
-        simpleSource.ack(ackRequest);
+
+        for (Offset offset : offsets) {
```
Comment (KeranYang) [inline examples/src/test/java/io/numaproj/numaflow/examples/source/simple/SimpleSourceTest.java:44]: please also update the testPending case.

File: src/main/java/io/numaproj/numaflow/sourcer/AckRequest.java [hunk 6]
```diff
@@ -1,14 +1,12 @@
...
 
-import java.util.List;
-
 /**
  * AckRequest request for acknowledging messages.
  */
 public interface AckRequest {
     /**
-     * @return the list of offsets to be acknowledged
+     * @return the offsets to be acknowledged
```
Comment (KeranYang) [inline src/main/java/io/numaproj/numaflow/sourcer/AckRequest.java:9]:
```suggestion
     * @return the offset to be acknowledged
```

File: src/main/java/io/numaproj/numaflow/sourcer/Service.java [hunk 7]
```diff
@@ -27,59 +25,112 @@ public Service(Sourcer sourcer) {
...
-                request.getRequest().getNumRecords(),
-                Duration.ofMillis(request.getRequest().getTimeoutInMs()));
-        this.sourcer.read(readRequest, outputObserver);
-        responseObserver.onCompleted();
+    public StreamObserver<SourceOuterClass.ReadRequest> readFn(final StreamObserver<SourceOuterClass.ReadResponse> responseObserver) {
+        return new StreamObserver<>() {
+            @Override
+            public void onNext(SourceOuterClass.ReadRequest request) {
+                // if the request is a handshake, send handshake response.
+                if (request.hasHandshake() && request.getHandshake().getSot()) {
```
Comment (KeranYang) [inline src/main/java/io/numaproj/numaflow/sourcer/Service.java:36]: Can we unit test this branch? Same for ack func below.

File: src/main/java/io/numaproj/numaflow/sourcer/SourcerTestKit.java [hunk 8]
```diff
@@ -124,33 +124,35 @@ public void sendReadRequest(ReadRequest request, OutputObserver observer) throws
...
-                public void onError(Throwable t) {
-                    future.completeExceptionally(t);
-                }
-
-                @Override
-                public void onCompleted() {
-                    future.complete(true);
-                }
-            });
+            StreamObserver<SourceOuterClass.ReadRequest> readRequestStreamObserver = sourceStub.readFn(
```
Comment (KeranYang) [inline src/main/java/io/numaproj/numaflow/sourcer/SourcerTestKit.java:127]: we can exclude SourceTestKit from codcov.

**Review by KeranYang — APPROVED** [review]

**Pull request merged**